### ISSUE TITLE
IndexOutOfBoundsException SpacedEditText.setSelection(SpacedEditText.java:82)

### ISSUE THREAD

**Issue by krawa** [firebase/FirebaseUI-Android#779]
I get a lot of reports with this crash.

firebase-ui-auth:2.0.1

Android 6.0
Lenovo VIBE K10 (K10a40)
Lenovo A6600 Plus (A6600a40) 

Android 5.0
Lenovo A5000 (A5000)

```
java.lang.IndexOutOfBoundsException: 
  at android.text.SpannableStringBuilder.checkRange(SpannableStringBuilder.java:1108)
  at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:680)
  at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:673)
  at android.text.Selection.setSelection(Selection.java:76)
  at android.text.Selection.setSelection(Selection.java:87)
  at android.widget.EditText.setSelection(EditText.java:115)
  at com.firebase.ui.auth.ui.phone.SpacedEditText.setSelection(SpacedEditText.java:82)
  at android.widget.EditText.setText(EditText.java:99)
  at com.firebase.ui.auth.ui.phone.SpacedEditText.setText(SpacedEditText.java:62)
  at android.widget.TextView.setText(TextView.java:4222)
  at com.firebase.ui.auth.ui.phone.SubmitConfirmationCodeFragment.setupConfirmationCodeEditText(SubmitConfirmationCodeFragment.java:192)
  at com.firebase.ui.auth.ui.phone.SubmitConfirmationCodeFragment.onCreateView(SubmitConfirmationCodeFragment.java:88)
  at android.support.v4.app.Fragment.performCreateView(Fragment.java:2192)
  at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:1299)
  at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManagerImpl.java:1528)
  at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:1595)
  at android.support.v4.app.BackStackRecord.executeOps(BackStackRecord.java:758)
  at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManagerImpl.java:2363)
  at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManagerImpl.java:2149)
  at android.support.v4.app.FragmentManagerImpl.optimizeAndExecuteOps(FragmentManagerImpl.java:2103)
  at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManagerImpl.java:2013)
  at android.support.v4.app.FragmentManagerImpl$1.run(FragmentManagerImpl.java:710)
  at android.os.Handler.handleCallback(Handler.java:815)
  at android.os.Handler.dispatchMessage(Handler.java:104)
  at android.os.Looper.loop(Looper.java:207)
  at android.app.ActivityThread.main(ActivityThread.java:5728)
  at java.lang.reflect.Method.invoke(Method.java:0)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:789)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:679)

```
Same crash i got in Digits

```

Caused by: java.lang.IndexOutOfBoundsException: 
  at android.text.SpannableStringBuilder.checkRange(SpannableStringBuilder.java:1108)
  at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:680)
  at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:673)
  at android.text.Selection.setSelection(Selection.java:76)
  at android.text.Selection.setSelection(Selection.java:87)
  at android.widget.EditText.setSelection(EditText.java:115)
  at com.digits.sdk.android.SpacedEditText.setSelection(SpacedEditText.java:87)
  at android.widget.EditText.setText(EditText.java:99)
  at com.digits.sdk.android.SpacedEditText.setText(SpacedEditText.java:68)
  at android.widget.TextView.setText(TextView.java:4222)
  at com.digits.sdk.android.ConfirmationCodeActivityDelegate.setUpEditText(ConfirmationCodeActivityDelegate.java:144)
  at com.digits.sdk.android.ConfirmationCodeActivityDelegate.init(ConfirmationCodeActivityDelegate.java:99)
  at com.digits.sdk.android.DigitsActivity.onCreate(DigitsActivity.java:42)
  at com.digits.sdk.android.ConfirmationCodeActivity.onCreate(ConfirmationCodeActivity.java:24)
  at android.app.Activity.performCreate(Activity.java:6309)
  at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1113)
  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2519)

```

**Comment by samtstern** [comment]
@ashwinraghav this was Digits code originally right?  We might have brought over a bug.

**Comment by ashwinraghav** [comment]
Yes. This code originally resided in digits. This is a long standing bug that we were never able to repro since we never got our hands on those devices. @krawa Are you able to identify a fix with the device you have?

**Comment by localhostEU** [comment]
Happens only on Lenovo devices.
Bug migrated from Digits.

If anyone have a Lenovo device - let's fix this!

**Comment by SUPERCILEX** [comment]
@krawa @localhostEU Please upgrade to FUI `v3.1.1` and send us new crash reports. We've added a bunch of logging that might hopefully help us debug the issue:
https://github.com/firebase/FirebaseUI-Android/blob/4b98a9853bf3cee9d0a32312ecf1897f740d1547/auth/src/main/java/com/firebase/ui/auth/ui/phone/SpacedEditText.java#L80-L84

**Comment by localhostEU** [comment]
@SUPERCILEX Great! Going to deploy in ~1h

**Comment by samtstern** [comment]
New logs in https://github.com/firebase/FirebaseUI-Android/issues/1129#issue-291836123

**Comment by samtstern** [comment]
Our call to `setSelection(index)` is actually equivalent to this:

https://developer.android.com/reference/android/text/Selection.html#setSelection(android.text.Spannable, int)
```
Selection.setSelection(getText(), index)
```

The docs say that method does this: "Move the cursor to offset index."

Here is our code:
```java
    @Override
    public void setSelection(int index) {
        // If the index is the leading edge, there are no spaces before it.
        // For all other cases, the index is preceded by index - 1 spaces.
        int spacesUpToIndex;
        if (index == 0) {
            spacesUpToIndex = 0;
        } else {
            spacesUpToIndex = index - 1;
        }

        try {
            super.setSelection(index + spacesUpToIndex);
        } catch (IndexOutOfBoundsException e) {
            // TODO remove once we figure out the bug.
            throw new IndexOutOfBoundsException(e.getMessage() +
                    ", requestedIndex=" + index +
                    ", spacesUpToIndex=" + spacesUpToIndex +
                    ", originalText=" + mOriginalText);
        }
    }
```

When index is `0` we do `setSelection(0)` and that's ok.  But if index is `x` we do `setSelection(2x -1)`.  Maybe we should just always do `setSelection(spacesUpToIndex)`?

**Comment by samtstern** [comment]
I also can't imagine how `requestedIndex` will ever get to 11 ....

**Comment by samtstern** [comment]
This has been fixed and released in version `3.2.2`.
